packages/compose/src/hooks/use-resize-observer/index.tsx

```diff
@@ -84,32 +84,21 @@ type ResizeElementProps = {
 };
 
 function ResizeElement( { onResize }: ResizeElementProps ) {
-	const resizeElementRef = useRef< HTMLDivElement >( null );
-	const resizeCallbackRef = useRef( onResize );
-
-	useLayoutEffect( () => {
-		resizeCallbackRef.current = onResize;
-	}, [ onResize ] );
-
-	useLayoutEffect( () => {
-		const resizeElement = resizeElementRef.current as HTMLDivElement;
-		const resizeObserver = new ResizeObserver( ( entries ) => {
-			for ( const entry of entries ) {
-				const newSize = extractSize( entry );
-				resizeCallbackRef.current( newSize );
-			}
-		} );
-
-		resizeObserver.observe( resizeElement );
-
-		return () => {
-			resizeObserver.unobserve( resizeElement );
-		};
-	}, [] );
+	const resizeCallbackEvent = useEvent( onResize );
+
+	const [ resizeElement, setResizeElement ] =
+		useState< HTMLDivElement | null >();
```

Comment:
This causes one extra render of the component that uses `useResizeObserver`. Because we store the element in state instead of a ref. The hook was better off when it was _not_ using the `useObserveElementSize` helper because there was one less `setState`.

I've been looking through `ResizeObserver` usages in Calypso and Gutenberg, and they _all_ either use a ref for the element, or could be fixed to use a ref instead of an element in state.

I'd say that refs are a "natural" place where to store DOM elements. They are almost never used by the render function itself, and almost always accessed only in effects. The `useObserveElementSize` hook should return a ref:
```js
const elRef = useObserveElementSize( onResize );
return <div ref={ elRef } />;
```
It can be tricky to catch when `elRef.current` changes to a different element, but most usages don't need it. They reliably pass the ref to an element that's always rendered.

Reply:
> It can be tricky to catch when `elRef.current` changes to a different element, but most usages don't need it. They reliably pass the ref to an element that's always rendered.

Should this be a scenario that we need to keep into account, though? Maybe we could use a callback ref to make sure we force a re-render only when needed?

Reply:
A ref-based hook would NOT work for my use-case. I need the element to be stateful. This hook needs to exist. Also, rendering is almost always extremely cheap.

That said, if you're still worried by the initial extra render, it'd be possible to make it compatible with refs, so we can get both things. Let me give it a go.

Reply:
Also, remember that, in composition, it's possible to go from minimal to more complex and not viceversa. This is what I discuss about here, though in relation to UI components: https://dio.la/article/the-everything-bagel-of-components#the-in-between-composability-is-unidirectional (composability is unidirectional).

I can make a utility that returns a ref from this utility, but I can't make a utility that takes an element from a utility that returns a ref, at least not without a hack that wastes cycles and memory. My utility here is the lowest common denominator of ResizeObserver abstractions, and now I'm gonna make it compatible with ref inputs too at the API level.

Reply:
See latest updates. Ref is supported now while still supporting the stateful case too. I updated the useResizeObserver hook too, and should now behave like before by using a ref. Win-win!

Reply:
> A ref-based hook would NOT work for my use-case.

The use case is the Tabs indication animation in #64926? I currently don't see the hook used there.

If a ref-based hook works for all use cases except one, then that one place should define its own custom hook. Instead of being the baseline for all others. What I see in the codebase is plenty of use cases for a hook that returns a ref.

> Should this be a scenario that we need to keep into account, though?

In theory, yes, in practice it doesn't need to be handled. For comparison, consider how a DOM listener would be added in an effect:
```js
const divRef = useRef(null);

useEffect(() => {
  const handleScroll = (event) => {
    console.log('Div scrolled!', event);
  };

  const div = divRef.current;
  if (div) {
    div.addEventListener('scroll', handleScroll);
  }

  return () => {
    if (div) {
      div.removeEventListener('scroll', handleScroll);
    }
  };
}, []);

return <div ref={divRef} />;
```
This is ChatGPT generated code and it's a common idiom. And yet it doesn't account for the case when the identity of the reffed `div` changes. And the baseline `useObserverElementSize` is used exactly the same way, it just creates a `ResizeObserver` instead of attaching an event listener.

If you need something more sophisticated, you should probably write your own hook from scratch.

Reply:
The example: https://github.com/WordPress/gutenberg/pull/64926/files#diff-64263a335082a1c717edd91225f49242f93a3b2625646af8af012dc96a6ff7a6R32

It goes through one layer of indirection that should be easy to figure out, but the point is that the resize observer needs to be updated when the active tab element changes (https://github.com/WordPress/gutenberg/pull/64926/files#diff-64263a335082a1c717edd91225f49242f93a3b2625646af8af012dc96a6ff7a6R33).

Reply:
I want to make extra emphasis on **fundamental**. I don't think utilities should cover every single use case, in fact I think they shouldn't. Your simplification of `useResizeObserver` shows why: YAGNI. Composition is king.

However, to achieve good composition, the primitive utilities do need to cover ALL **fundamental** use cases. Here are the ones that my utility covers:

- Track an element:
  - Statefully (update ResizeObserver on element change).
  - Non-statefully ("fire and forget" style, ref).
- Alert when size changes (callback).
- Pass ResizeObserver entries (passed to the callback).
- Pass configuration to ResizeObserver.

These are all of the fundamental use-cases of a ResizeObserver abstraction in a React context. I don't think config has ever been passed. Does that mean we shouldn't allow it as input? Does that mean that, if someone needs to pass config, they should create a local fork of everything just to have support for it?

With some rare exceptions, covering all **fundamental** use-cases is what reduces duplication and enables composition. Specific use-cases can use specific utilities, which are nothing but the composition of primitive utilities. And that's only possible when the primitives are minimal and comprehensive to the extreme. Thanks for coming to my TED talk.

Reply:
What I worry about is that the generic hook is a **zero-cost abstraction**. If I can write my own hook that avoids all the extra work and is faster, then the abstraction is not very useful.

I don't warn the `useResizeObserver` trigger extra rerenders that are not needed. A pattern preview already has three resize observers, i.e., 6 initial renders. I don't want that to be 9 🙂 If there is a big list of pattern previews, then you have a UI that renders all the time. That's at least very annoying to debug. You set breakpoints or logpoints and then most of them are bogus.

Reply:
Can you give an example of a hook that does less work than this one when you pass it a ref?

I think you might have misunderstood what my hook does, because it has the absolute lowest cost you can get for each of the input types (ref or stateful element). I'm going to update the example as I just realized it's not updated with ref support and I agree for most cases it should be the default use.

Reply:
I updated the JSDoc example, let me know if it clarifies things!